C Error: Can't return mutable sumtype from inside match

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.4 f19197f
**OS:**  Arch Linux

**What did you do?**

```v
module main

type Sum = Struct | int

struct Struct {
mut:
	value int
}

fn sum(mut s Sum) Sum {
	match mut s {
		Struct {
			return s
		}
		else {
			return s
		}
	}
}

fn main() {
	mut s := Sum(Struct{ value: 42 })
	s = sum(mut s)
	dump(s)
}
```


**What did you expect to see?**
`sum` returning the wrapped `Struct` without error and dumping `value: 42`
 
**What did you see instead?**
`tcc` produces a meaningless error message
```
==================
/tmp/v_1000/01.9250439459571540933.tmp.c:11121: error: '{' expected (got ";")
...
==================
```

`clang` has a better message
```
/tmp/v_1000/01.9069514595829982444.tmp.c:2043:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_1000/01.9069514595829982444.tmp.c:11093:13: error: initializing 'main__Sum' (aka 'struct main__Sum') with an expression of incompatible type 'main__Sum *' (aka 'struct main__Sum *'); dereference with *
                main__Sum _t1 = s;
                          ^     ~
                                *
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2421,9 +2421,9 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 	got_styp := g.typ(got_type)
 	if expected_type != ast.void_type {
 		unwrapped_expected_type := g.unwrap_generic(expected_type)
-		unwrapped_got_type := g.unwrap_generic(got_type)
 		unwrapped_exp_sym := g.table.sym(unwrapped_expected_type)
-		unwrapped_got_sym := g.table.sym(unwrapped_got_type)
+		mut unwrapped_got_type := g.unwrap_generic(got_type)
+		mut unwrapped_got_sym := g.table.sym(unwrapped_got_type)
 
 		expected_deref_type := if expected_is_ptr {
 			unwrapped_expected_type.deref()
@@ -2445,11 +2445,15 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 					is_already_sum_type = true
 				}
 			}
-			if is_already_sum_type {
+			if is_already_sum_type && !g.inside_return {
 				// Don't create a new sum type wrapper if there is already one
 				g.prevent_sum_type_unwrapping_once = true
 				g.expr(expr)
 			} else {
+				if mut unwrapped_got_sym.info is ast.Aggregate {
+					unwrapped_got_type = unwrapped_got_sym.info.types[g.aggregate_type_idx]
+					unwrapped_got_sym = g.table.sym(unwrapped_got_type)
+				}
 				g.get_sumtype_casting_fn(unwrapped_got_type, unwrapped_expected_type)
 				fname := '${unwrapped_got_sym.cname}_to_sumtype_$unwrapped_exp_sym.cname'
 				g.call_cfn_for_casting_expr(fname, expr, expected_is_ptr, unwrapped_exp_sym.cname,

--- a/vlib/v/tests/match_sumtype_var_return_sumtype_test.v
+++ b/vlib/v/tests/match_sumtype_var_return_sumtype_test.v
@@ -1,0 +1,30 @@
+module main
+
+type Sum = Struct | int
+
+struct Struct {
+mut:
+	value int
+}
+
+fn sum(mut s Sum) Sum {
+	match mut s {
+		Struct {
+			return s
+		}
+		else {
+			return s
+		}
+	}
+}
+
+fn test_match_sumtype_var_return_sumtype() {
+	mut s := Sum(Struct{
+		value: 42
+	})
+	s = sum(mut s)
+	dump(s)
+	ret := '$s'
+	assert ret.contains('Sum(Struct{')
+	assert ret.contains('value: 42')
+}

